File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Trpak/Archive/TrpfsLoader.cs
using MiniToolbox.Core.Archive;
using System.Runtime.InteropServices;
using MiniToolbox.Trpak.Flatbuffers.TR.ResourceDictionary;
using MiniToolbox.Core.Utils;

namespace MiniToolbox.Trpak.Archive
{
    /// <summary>
    /// Loads files from a TRPFD/TRPFS archive pair (Pokémon Scarlet/Violet, Legends Arceus).
    /// data.trpfd = FlatBuffer file descriptor (hashes → pack names/indices).
    /// data.trpfs = ONEFILE header + FlatBuffer FileSystem (hashes → offsets) + packed data.
    /// </summary>
    public class TrpfsLoader
    {
        private readonly CustomFileDescriptor _fd;
        private readonly Flatbuffers.TR.ResourceDictionary.FileSystem _fs;
        private readonly string _trpfsPath;
        private readonly TrpakHashCache _hashCache;

        // Pack cache: packHash → deserialized PackedArchive
        private readonly Dictionary<ulong, PackedArchive> _packCache = new();

        // O(1) lookup indexes — replaces Array.IndexOf which was O(n)
        private readonly Dictionary<ulong, int> _fdHashIndex;
        private readonly Dictionary<ulong, int> _fsHashIndex;
        private readonly Dictionary<ulong, int>? _fdUnusedHashIndex;

        public TrpfsLoader(string arcDirectory, TrpakHashCache? hashCache = null)
        {
            string trpfdPath = Path.Combine(arcDirectory, "data.trpfd");
            _trpfsPath = Path.Combine(arcDirectory, "data.trpfs");

            if (!File.Exists(trpfdPath))
                throw new FileNotFoundException("data.trpfd not found.", trpfdPath);
            if (!File.Exists(_trpfsPath))
                throw new FileNotFoundException("data.trpfs not found.", _trpfsPath);

            _fd = FlatBufferConverter.DeserializeFrom<CustomFileDescriptor>(trpfdPath);
            _fs = ReadFileSystem(_trpfsPath);
            _hashCache = hashCache ?? new TrpakHashCache();

            // Build O(1) lookup indexes
            _fdHashIndex = BuildHashIndex(_fd.FileHashes);
            _fsHashIndex = BuildHashIndex(_fs.FileHashes);
            if (_fd.UnusedHashes != null && _fd.UnusedHashes.Length > 0)
                _fdUnusedHashIndex = BuildHashIndex(_fd.UnusedHashes);
        }

        private static Dictionary<ulong, int> BuildHashIndex(ulong[]? hashes)
        {
            var dict = new Dictionary<ulong, int>();
            if (hashes == null) return dict;
            for (int i = 0; i < hashes.Length; i++)
                dict.TryAdd(hashes[i], i);
            return dict;
        }

        /// <summary>Number of files in the descriptor.</summary>
        public int FileCount => _fd.FileHashes?.Length ?? 0;

        /// <summary>Raw file hashes from the descriptor.</summary>
        public ulong[] FileHashes => _fd.FileHashes ?? Array.Empty<ulong>();

        /// <summary>All pack names registered in the descriptor.</summary>
        public IReadOnlyList<string> PackNames => _fd.PackNames;

        /// <summary>Get file info (pack index) for a file by its index in FileHashes.</summary>
        public Flatbuffers.TR.ResourceDictionary.FileInfo? GetFileInfo(int fileIndex)
        {
            if (_fd.FileInfo != null && fileIndex >= 0 && fileIndex < _fd.FileInfo.Length)
                return _fd.FileInfo[fileIndex];
            return null;
        }

        /// <summary>
        /// Extract a file by its romfs-relative path (e.g. "pokemon/pokemon_model/pm0025/pm0025_00/mdl/pm0025_00.trmdl").
        /// </summary>
        public byte[]? ExtractFile(string romfsRelativePath)
        {
            string normalized = NormalizePath(romfsRelativePath);
            ulong hash = FnvHash.Hash(normalized);
            return ExtractFile(hash);
        }

        /// <summary>
        /// Extract a file by its FNV hash.
        /// </summary>
        public byte[]? ExtractFile(ulong fileHash)
        {
            if (!TryResolvePackInfo(fileHash, out string packName, out long packSize))
                return null;

            ulong packHash = FnvHash.Hash(packName);
            if (!TryGetPack(packHash, packSize, out var pack))
                return null;

            int entryIndex = FindEntryIndex(pack, fileHash);
            if (entryIndex < 0)
                return null;

            var entry = pack.FileEntry[entryIndex];
            byte[] buffer = entry.FileBuffer;

            // EncryptionType != -1 means Oodle-compressed
            if (entry.EncryptionType != -1)
            {
                var decompressed = OodleDecompressor.Decompress(buffer, (int)entry.FileSize);
                if (decompressed == null)
                    throw new InvalidDataException($"Oodle decompression failed for hash 0x{fileHash:X16}");
                buffer = decompressed;
            }

            return buffer;
        }

        /// <summary>
        /// Find all file hashes whose human-readable names match a predicate.
        /// Requires a populated hash cache.
        /// </summary>
        public IEnumerable<(ulong hash, string name)> FindFiles(Func<string, bool> namePredicate)
        {
            if (_fd.FileHashes == null) yield break;

            foreach (ulong hash in _fd.FileHashes)
            {
                string? name = _hashCache.GetName(hash);
                if (name != null && namePredicate(name))
                    yield return (hash, name);
            }
        }

        /// <summary>
        /// Find all files with a given extension (e.g. ".trmdl").
        /// </summary>
        public IEnumerable<(ulong hash, string name)> FindFilesByExtension(string extension)
        {
            return FindFiles(name => name.EndsWith(extension, StringComparison.OrdinalIgnoreCase));
        }

        /// <summary>
        /// Get the pack index for a given file hash.
        /// </summary>
        public int GetPackIndexForFile(ulong fileHash)
        {
            if (_fd.FileHashes == null || _fd.FileInfo == null) return -1;
            if (!_fdHashIndex.TryGetValue(fileHash, out int idx)) return -1;
            return (int)_fd.FileInfo[idx].PackIndex;
        }

        /// <summary>
        /// Get all file hashes that belong to a given pack index.
        /// </summary>
        public List<ulong> GetFilesInPack(int packIndex)
        {
            var result = new List<ulong>();
            if (_fd.FileHashes == null || _fd.FileInfo == null) return result;
            for (int i = 0; i < _fd.FileHashes.Length; i++)
            {
                if ((int)_fd.FileInfo[i].PackIndex == packIndex)
                    result.Add(_fd.FileHashes[i]);
            }
            return result;
        }

        /// <summary>
        /// Get the pack name for a given pack index.
        /// </summary>
        public string? GetPackName(int packIndex)
        {
            if (_fd.PackNames != null && packIndex >= 0 && packIndex < _fd.PackNames.Length)
                return _fd.PackNames[packIndex];
            return null;
        }

        #region Private helpers

        private static string NormalizePath(string path)
        {
            path = (path ?? string.Empty).Replace('\\', '/').Trim();
            if (path.StartsWith("romfs://", StringComparison.OrdinalIgnoreCase))
                path = path["romfs://".Length..];
            if (path.StartsWith("trpfs://", StringComparison.OrdinalIgnoreCase))
                path = path["trpfs://".Length..];
            return path.TrimStart('/');
        }

        private static Flatbuffers.TR.ResourceDictionary.FileSystem ReadFileSystem(string trpfsPath)
        {
            using var br = new BinaryReader(File.OpenRead(trpfsPath));

            // OneFileHeader: 8 bytes magic + 8 bytes offset
            br.ReadUInt64(); // magic
            long fsOffset = br.ReadInt64();

            br.BaseStream.Position = fsOffset;
            byte[] fsData = br.ReadBytes((int)(br.BaseStream.Length - fsOffset));

            return FlatBufferConverter.DeserializeFrom<Flatbuffers.TR.ResourceDictionary.FileSystem>(fsData);
        }

        private bool TryResolvePackInfo(ulong fileHash, out string packName, out long packSize)
        {
            packName = string.Empty;
            packSize = 0;

            if (_fd.FileHashes == null || _fd.FileInfo == null || _fd.PackNames == null || _fd.PackInfo == null)
                return false;

            if (_fdHashIndex.TryGetValue(fileHash, out int idx))
            {
                ulong packIndex = _fd.FileInfo[idx].PackIndex;
                if (packIndex < (ulong)_fd.PackNames.Length && packIndex < (ulong)_fd.PackInfo.Length)
                {
                    packName = _fd.PackNames[packIndex];
                    packSize = checked((long)_fd.PackInfo[packIndex].FileSize);
                    return !string.IsNullOrWhiteSpace(packName);
                }
                return false;
            }

            // Check unused files
            if (_fdUnusedHashIndex != null && _fd.UnusedFileInfo != null)
            {
                if (_fdUnusedHashIndex.TryGetValue(fileHash, out int unusedIdx) && unusedIdx < _fd.UnusedFileInfo.Length)
                {
                    ulong packIndex = _fd.UnusedFileInfo[unusedIdx].PackIndex;
                    if (packIndex < (ulong)_fd.PackNames.Length && packIndex < (ulong)_fd.PackInfo.Length)
                    {
                        packName = _fd.PackNames[packIndex];
                        packSize = checked((long)_fd.PackInfo[packIndex].FileSize);
                        return !string.IsNullOrWhiteSpace(packName);
                    }
                }
            }

            return false;
        }

        private const int MaxPackCacheSize = 10;

        private bool TryGetPack(ulong packHash, long packSize, out PackedArchive pack)
        {
            if (_packCache.TryGetValue(packHash, out pack!))
                return true;

            if (!_fsHashIndex.TryGetValue(packHash, out int fileIndex))
            {
                pack = null!;
                return false;
            }

            // Read raw pack bytes from TRPFS (streaming, not buffered)
            using var br = new BinaryReader(File.OpenRead(_trpfsPath));
            br.BaseStream.Position = (long)_fs.FileOffsets[fileIndex];
            byte[] packBytes = br.ReadBytes((int)packSize);

            pack = FlatBufferConverter.DeserializeFrom<PackedArchive>(packBytes);

            // Evict oldest if cache is full
            if (_packCache.Count >= MaxPackCacheSize)
                _packCache.Remove(_packCache.Keys.First());

            _packCache[packHash] = pack;
            return true;
        }

        /// <summary>Clear the pack cache to free memory.</summary>
        public void ClearPackCache() => _packCache.Clear();

        private static int FindEntryIndex(PackedArchive pack, ulong fileHash)
        {
            if (pack.FileHashes == null) return -1;
            return Array.IndexOf(pack.FileHashes, fileHash);
        }

        #endregion
    }
}

File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Api/src/lib/Archive/TrpfsLoader.ts
/**
 * TRPFS/TRPFD Archive Loader
 * Ported from C# TrpfsLoader.cs
 * 
 * Loads files from a TRPFD/TRPFS archive pair (Pokémon Scarlet/Violet, Legends Arceus).
 * data.trpfd = FlatBuffer file descriptor (hashes → pack names/indices).
 * data.trpfs = ONEFILE header + FlatBuffer FileSystem (hashes → offsets) + packed data.
 */

import * as fs from 'fs';
import * as path from 'path';
import {
    FileDescriptor,
    CustomFileDescriptor,
    FileSystem,
    PackedArchive,
    PackedFile
} from '../Flatbuffers/TR/ResourceDictionary/index.js';
import { FlatBufferConverter } from '../Utils/FlatBufferConverter.js';
import { TrpakHashCache, FnvHash } from './TrpakTypes.js';
import { OodleDecompressor } from './Decompressors.js';

export class TrpfsLoader {
    private _fd: CustomFileDescriptor;
    private _fs: FileSystem;
    private _trpfsPath: string;
    private _hashCache: TrpakHashCache;

    // Pack cache: packHash → deserialized PackedArchive
    private _packCache: Map<bigint, PackedArchive> = new Map();

    constructor(arcDirectory: string, hashCache?: TrpakHashCache) {
        const trpfdPath = path.join(arcDirectory, 'data.trpfd');
        this._trpfsPath = path.join(arcDirectory, 'data.trpfs');

        if (!fs.existsSync(trpfdPath)) {
            throw new Error(`data.trpfd not found: ${trpfdPath}`);
        }
        if (!fs.existsSync(this._trpfsPath)) {
            throw new Error(`data.trpfs not found: ${this._trpfsPath}`);
        }

        this._fd = FlatBufferConverter.DeserializeFrom(trpfdPath, CustomFileDescriptor);
        this._fs = this.ReadFileSystem(this._trpfsPath);
        this._hashCache = hashCache ?? new TrpakHashCache();
    }

    /** Number of files in the descriptor. */
    get FileCount(): number {
        return this._fd.FileHashes?.length ?? 0;
    }

    /** All pack names registered in the descriptor. */
    get PackNames(): readonly string[] {
        return this._fd.PackNames;
    }

    /**
     * Extract a file by its romfs-relative path (e.g. "pokemon/pokemon_model/pm0025/pm0025_00/mdl/pm0025_00.trmdl").
     */
    ExtractFile(romfsRelativePath: string): Buffer | null {
        const normalized = this.NormalizePath(romfsRelativePath);
        const hash = FnvHash.Hash(normalized);
        return this.ExtractFileByHash(hash);
    }

    /**
     * Extract a file by its FNV hash.
     */
    ExtractFileByHash(fileHash: bigint): Buffer | null {
        const packInfo = this.TryResolvePackInfo(fileHash);
        if (!packInfo) return null;

        const { packName, packSize } = packInfo;
        const packHash = FnvHash.Hash(packName);
        const pack = this.TryGetPack(packHash, packSize);
        if (!pack) return null;

        const entryIndex = this.FindEntryIndex(pack, fileHash);
        if (entryIndex < 0) return null;

        const entry = pack.FileEntry[entryIndex];
        let buffer = entry.FileBuffer;

        // EncryptionType != -1 means Oodle-compressed
        if (entry.EncryptionType !== -1) {
            const decompressed = OodleDecompressor.Decompress(buffer, Number(entry.FileSize));
            if (decompressed === null) {
                throw new Error(`Oodle decompression failed for hash 0x${fileHash.toString(16).padStart(16, '0')}`);
            }
            buffer = decompressed;
        }

        return buffer;
    }

    /**
     * Find all file hashes whose human-readable names match a predicate.
     * Requires a populated hash cache.
     */
    *FindFiles(namePredicate: (name: string) => boolean): IterableIterator<[bigint, string]> {
        if (!this._fd.FileHashes) return;

        for (const hash of this._fd.FileHashes) {
            const name = this._hashCache.GetName(hash);
            if (name !== null && namePredicate(name)) {
                yield [hash, name];
            }
        }
    }

    /**
     * Find all files with a given extension (e.g. ".trmdl").
     */
    *FindFilesByExtension(extension: string): IterableIterator<[bigint, string]> {
        const ext = extension.toLowerCase();
        yield* this.FindFiles(name => name.toLowerCase().endsWith(ext));
    }

    private NormalizePath(inputPath: string): string {
        let p = (inputPath ?? '').replace(/\\/g, '/').trim();

        if (p.toLowerCase().startsWith('romfs://')) {
            p = p.slice('romfs://'.length);
        }
        if (p.toLowerCase().startsWith('trpfs://')) {
            p = p.slice('trpfs://'.length);
        }

        return p.replace(/^\/+/, '');
    }

    private ReadFileSystem(trpfsPath: string): FileSystem {
        const fd = fs.openSync(trpfsPath, 'r');
        try {
            // OneFileHeader: 8 bytes magic + 8 bytes offset
            const header = Buffer.alloc(16);
            fs.readSync(fd, header, 0, 16, 0);
            const fsOffset = Number(header.readBigInt64LE(8));

            // Read remaining bytes from fsOffset to end of file
            const stat = fs.fstatSync(fd);
            const fsSize = Number(stat.size) - fsOffset;
            const fsData = Buffer.alloc(fsSize);
            fs.readSync(fd, fsData, 0, fsSize, fsOffset);

            return FlatBufferConverter.DeserializeFrom(fsData, FileSystem) as FileSystem;
        } finally {
            fs.closeSync(fd);
        }
    }

    private TryResolvePackInfo(fileHash: bigint): { packName: string; packSize: bigint } | null {
        if (!this._fd.FileHashes || !this._fd.FileInfo || !this._fd.PackNames || !this._fd.PackInfo) {
            return null;
        }

        let idx = this._fd.FileHashes.indexOf(fileHash);
        if (idx >= 0) {
            const packIndex = this._fd.FileInfo[idx].PackIndex;
            if (packIndex < BigInt(this._fd.PackNames.length) && packIndex < BigInt(this._fd.PackInfo.length)) {
                const packName = this._fd.PackNames[Number(packIndex)];
                const packSize = this._fd.PackInfo[Number(packIndex)].FileSize;
                if (packName && packName.trim().length > 0) {
                    return { packName, packSize };
                }
            }
            return null;
        }

        // Check unused files
        if (this._fd.UnusedHashes && this._fd.UnusedFileInfo) {
            idx = this._fd.UnusedHashes.indexOf(fileHash);
            if (idx >= 0 && idx < this._fd.UnusedFileInfo.length) {
                const packIndex = this._fd.UnusedFileInfo[idx].PackIndex;
                if (packIndex < BigInt(this._fd.PackNames.length) && packIndex < BigInt(this._fd.PackInfo.length)) {
                    const packName = this._fd.PackNames[Number(packIndex)];
                    const packSize = this._fd.PackInfo[Number(packIndex)].FileSize;
                    if (packName && packName.trim().length > 0) {
                        return { packName, packSize };
                    }
                }
            }
        }

        return null;
    }

    private TryGetPack(packHash: bigint, packSize: bigint): PackedArchive | null {
        if (this._packCache.has(packHash)) {
            return this._packCache.get(packHash)!;
        }

        const fileIndex = this._fs.FileHashes.indexOf(packHash);
        if (fileIndex < 0) {
            return null;
        }

        // Read raw pack bytes from TRPFS using partial read (file can be >2GB)
        const packOffset = Number(this._fs.FileOffsets[fileIndex]);
        const packLen = Number(packSize);
        const packBytes = Buffer.alloc(packLen);
        const fd = fs.openSync(this._trpfsPath, 'r');
        try {
            fs.readSync(fd, packBytes, 0, packLen, packOffset);
        } finally {
            fs.closeSync(fd);
        }

        const pack = FlatBufferConverter.DeserializeFrom(packBytes, PackedArchive) as PackedArchive;
        this._packCache.set(packHash, pack);
        return pack;
    }

    private FindEntryIndex(pack: PackedArchive, fileHash: bigint): number {
        if (!pack.FileHashes) return -1;
        return pack.FileHashes.indexOf(fileHash);
    }
}
